subiquity/ui/views/snaplist.py

```diff
@@ -0,0 +1,367 @@
+# Copyright 2015 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from urwid import (
+    BOX,
+    CheckBox,
+    LineBox,
+    RadioButton,
+    SelectableIcon,
+    Text,
+    Widget,
+    WidgetWrap,
+    )
+
+from subiquitycore.ui.buttons import ok_btn, cancel_btn, other_btn
+from subiquitycore.ui.container import Columns, ListBox, Pile
+from subiquitycore.ui.utils import button_pile, Color, Padding, screen
+from subiquitycore.view import BaseView
+
+from subiquity.models.filesystem import humanize_size
+from subiquity.models.snaplist import SnapSelection
+from subiquity.ui.spinner import Spinner
+
+log = logging.getLogger("subiquity.views.snaplist")
+
+
+class StarCheckBox(CheckBox):
+    states = {
+        True: SelectableIcon(" *"),
+        False: SelectableIcon("  "),
+        }
+    reserve_columns = 3
+
+
+class StarRadioButton(RadioButton):
+    states = {
+        True: SelectableIcon(" *"),
+        False: SelectableIcon("  "),
+        }
+    reserve_columns = 3
+
+
+class NoTabCyclingListBox(ListBox):
+    # Carefully disable the TabCycling parts of our ListBox (but keep the scrollbar!)
+
+    def keypress(self, size, key):
+        if not key.startswith("enter") and self._command_map[key] in ('next selectable', 'prev selectable'):
+            return key
+        else:
+            return super().keypress(size, key)
+
+    def _select_first_selectable(self):
+        return
+
+    def _select_last_selectable(self):
+        return
+
+
+class SnapInfoView(Widget):
+
+    # This is mostly like a Pile but it tries to be a bit smart about
+    # how to distribute space between the description and channel list
+    # (which can both be arbitrarily long or short). If both are long,
+    # the channel list is given a third of the space. If there is
+    # space for both, they are packed into the upper part of the view.
+
+    _selectable = True
+    _sizing = frozenset([BOX])
+
+    def __init__(self, parent, snap, cur_channel):
+        self.parent = parent
+        self.snap = snap
+        self.channels = []
+        self.needs_focus = True
+
+        channel_width = max(len(csi.channel_name) for csi in snap.channels) \
+          + StarRadioButton.reserve_columns + 1
```

Comment:
Generally you can wrap a new line by putting the expression in parens:

(max() for csi in snap + 
   Button.cols +1)

Reply:
yeah that is usually a better idea, fixed.